MCP/0035 Multilingual support of Modelica (#2956)
modelica/ModelicaSpecification: RationaleMCP/0035/SpecificationText.md

"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
```
All general terms in the header should be replaced by specific information.

In the body, directly following the header, for each descriptive text there is an entry of the form
```
#: <file name>:<line number>
msgctxt "<context identifier>"
msgid "<descriptive text>"
msgstr "<translation>"
```
Only the keywords `msgctxt`, `msgid` and `msgstr` are used.

At first there can be an optional comment describing the location (file name and line number) of the text to translate. Multiple occurences of the same string can be listed here, separated by space.

Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes belong to the enclosing full class definition.

After the `msgid` keyword the text string which shall be translated follows. It should contain the original string from the Modelica text representation. 

> Since in Modelica control sequences also start with a backslash and another backslash is used to use sequences literally or to hide double quotes, no change is required here. 

Reviewer: I am trying to understand the relation between what should be specified specifically for the proposed used in Modelica library translations, and what should be explained just as examples of how the PO format can be utilized.

Shouldn't the emphasis here be on the actual string data content, and not which of all the available PO file mechanisms that are used to encode it in the file?

In that case, shouldn't it be sufficient to say that what goes into the `msgid` is the value of the Modelica string literal?  This would have the consequence that there is no distinction between a literal newline and `\n` in the original Modelica string literal; they both become a newline in the value of the string, and then it is up to the PO file editor to choose a quoting mechanism.

Consequently, the Modelica string literal `"a\\nb"` is evaluated to a string of length three, and one of the many alternatives for expressing this in a PO file would be this:
```
#, c-format
msgid "a\\nb"
```

From a translation point of view, I only see benefits of not making a distinction between a literal newline and `\n` in the original Modelica string literals.

Author: Sure, it could be sufficient to define msgid as string containinng the string literal from Modelica text. But multiline strings have to be converted to multiple strings ending with `\n`, see below. Whereby a `\n` in the middle of the string remain unchanged.

Btw.`"a\\nb"` ist a string of length four.

Reviewer: Sorry, I don't know what happened to the escapes in my comment above.  I'll try to edit the comment, changing `\\n` → `\n`.

Reviewer: > Sure, it could be sufficient to define msgid as string containinng the string literal from Modelica text. But multiline strings have to be converted to multiple strings ending with `\n`, see below. Whereby a `\n` in the middle of the string remain unchanged.

The point I am trying to make is that the Modelica string _literal_ should be evaluated according to Modelica string literal semantics, resulting in a string _value_.  Then, this string _value_ is what gets encoded in the gettext files using one of the available format string alternatives.  I suppose that the different format string alternatives are rather similar when it comes to escape sequences, so the Modelica string literal `"a\nb"` is likely to actually look the same in most or all format string variants.

What I don't see clearly is what the choice of format string would mean for placeholders in Modelica strings, both today and in the future.  At the moment, Modelica has placeholders described for `Text` and `Figure` annotations:
- https://specification.modelica.org/master/annotations.html#text
- https://specification.modelica.org/master/annotations.html#variable-replacements

For a Modelica string such as `"Dimensions: %{width} by %{height}"`, what difference would the choice of gettext format string make?  If gettext's `c-format` is used, would it be a problem that the string as given would be a malformed C format string?

Author: Modelica strings are not evaluated, they are used as they are, with all patterns unchanged. The translation has to reproduce the patterns to get the same information at the end. gettexxt itself does not know anything about the sematics of patterns, they are seen as plain text there.

Reviewer 2: And note that gettext's c-format handling is optional in the translation tools, so I don't see any problem with it here.
This is contrast to e.g., QT's translation-handling that always seems to be active.

Reviewer: I guess we can consider this conversation resolved by now?

But Modelica allows strings to go over more than one line, gettext does not.
Therefore, Modelica strings must be split into consecutive strings at the line breaks, and the line breaks must be encoded with a "\n" at the end of each string, e.g.
```
"A
B"
```
is converted to
```
"A\n"
"B"

> ```

Reviewer: Adding to my comment above, I think that the kind of splitting into separate strings as proposed here would be catastrophic from a translation point of view.  Literal line breaks are often inserted into long strings for hard line wrapping of a message to fit a certain column width.  The splitting carelessly breaks down sentences into fragments that may not translate sensibly on their own, and trying to preserve the style of hard line wrapping to the same column width would be an insane task.  (That said, I think hard line wraps in message strings are a bad idea, but it is not our business to argue over such stylistic matters here.)

Reviewer 2: As I understand: it's **not separate** strings, but one conceptual string running over several lines.
I would assume the tools for handling translations don't care whether the string is split over several lines or not - but only care about the "\n".

Author: This splitting of strings serves only to support existing translation tools which follow the gettext format specification which does not support strings going over multiple lines. The tools themselves will join the lines and present them to the user as one text to translate.

Reviewer: OK, thanks for the explanation.  I now found the corresponding passage in the gettext documentation.

In that case, I find this another example of something that shouldn't be dictated normatively by the Modelica specification.  A non-normative comment mentioning the possibility of splitting over multiple lines should be sufficient.  After all, with tools used to edit these files, why should we even care whether the strings are split over several lines or not?

Author: @HansOlsson has put the line breaking stuff in non normative text. This is only to know how to generate translation files from Modelica and how to read in multiple consecutive strings from a translation file into a Modelica tool.

Reviewer 2: To clarify I put line break->"\n" in normative text, but the actual splitting of text for gettext in the non-normative part.

Reviewer 2: I added 'for gettext' to make that clearer.

Reviewer 2: I believe this is now clear, and we can start the next step in the process. @henrikt-ma  agreed?

Reviewer: I'm afraid that my biggest problem is insufficient understanding of the relation between the Modelica, its specification, and the gettext specification.  I still have a feeling that the proposed Modelica specification text is too much a repetition of the gettext specification, instead of putting emphasis on how the gettext specification should be applied in the Modelica context.

One detail I don't see answered at the moment is the role of the format string, and I don't understand whether such a statement should be normative or if a non-normative remark would be sufficient.

Reviewer 2: I don't understand what the _format string_ is intended to mean.
The basic idea is to take the strings from the Modelica code and add them almost verbatim to the gettext-file (verbatim except for the line-break handling), with context given by class-name etc.

Reviewer: I'm referring to section 15.3 _The Translator's View_ in the gettext specification.  The section begins with:
> The translator works exactly as in the C language case. The only difference is that when translating format strings, she has to be aware of the language’s particular syntax for positional arguments in format strings.

To me this means that it doesn't matter what the Modelica string input syntax looks like; the strings to be translated should be evaluated (removing the difference between literal newlines and `\n`), and the string data should be expressed using C string syntax (encoding all newlines as `\n`), with some choice of positional argument syntax on top of that.

The least we could do is to comment on the role of the gettext format string syntax in the context of Modelica translations.  A simple statement like this could be enough:
> No other gettext format string syntax than `no-c-format` may be specified, and the actual semantics of a string depends on the Modelica context.  For example, `%name` has special meaning in certain Modelica contexts and must then be treated similar to positional arguments in the standardized ways of expressing gettext format strings.

Reviewer: Actually, the fact that we have special meaning assigned to things like `%class` in some context means that `no-c-format` should actually be required wherever such constructs may appear.  To make things simpler, I'd suggest making `no-c-format` mandatory for all strings until we see a need to allow any other format.

Author: We should always put the comment `# no-c-format` in the line before msgid to avoid missinterpretation by gettext tools.

Author: see the change above.

# 13.6 Multilingual Descriptions

[*Descriptive texts in a model or library are usually formulated in English. To support users who are not so familiar with this, a Modelica tool can support the translation of these texts into any language. This should be done using external files, so that no modification of the actual Modelica text is required.*]

The files to support translation must be provided by the developer of the library. They must be stored in a subdirectory of the Resources directory of the library with the name `Language`.

Two kind of files have to be provided:

> * a template file `<LibraryName>.pot`. 

Reviewer 3: ```suggestion
* a template file `<LibraryName>.pot` (Portable Object Template). 
```

It contains all necessary information to translate all descriptions, but no translations. The pattern `<LibraryName>` denotes the toplevel class name of the library.

> * one file for each supported language with the name `<LibraryName>.<language>.po`. This file is a copy of the associated template file, but extended with the translations in the specified language. The pattern `<language>` stands for the ISO 639-1 language code, e.g. de or en.

Reviewer: Is it a deliberate decision to not support the binary _.mo_ file format? If so, I think it would be wise to state it clearly.

Reviewer 2: I would say it is a deliberate decision, since we don't want to standardize the implementation (especially not the gettext-specific hashing).

Reviewer: Right, let's add a normative statement about that, and see if it causes any reactions under the MCP evaluation.  Opening up for allowing the _.mo_ format later would of course be easy if we'd ever change our mind (for example, it it turns out that all tools are using the LGPL getttext library anyway in the end).

Reviewer: By the way, the way I imagine the Modelica library _.po_ files would be used in System Modeler is that the tool has a cache of _.mo_ files that are created when the library is loaded (the gettext _msgfmt_ program (GPL) can do this), so that the translated strings can then be accessed efficiently using the widely available _libintl_ (LGPL) library.  If anyone sees any catch with this approach, please bring it to light.

Reviewer 2: To me creating a binary file is something that tools may optionally add (in that format or some other one), but the specification standardizes on the ".po" file so that all tools can translate the library in the same way.

Reviewer: > To me creating a binary file is something that tools may optionally add (in that format or some other one), but the specification standardizes on the ".po" file so that all tools can translate the library in the same way.

I agree that this is the right way to go, which is why I am suggesting to add a non-normative comment to clarify that a library cannot be distributed with _.mo_-files as a substitute for the _.po_-files.

Reviewer 2: I added a generalization of that as a normative text; https://github.com/modelica/ModelicaSpecification/pull/2956#discussion_r983395460_ - resolving this to continue the discussion there.

> * one file for each supported language with the name `<LibraryName>.<language>.po`. This file is a copy of the associated template file, but extended with the translations in the specified language. The pattern `<language>` stands for the ISO 639-1 language code, e.g. de or en.

Reviewer 3: ```suggestion
* one file for each supported language with the name `<LibraryName>.<language>.po` (Portable Object). This file is a copy of the associated template file, but extended with the translations in the specified language. The pattern `<language>` stands for the ISO 639-1 language code, e.g., de or sv.
```

```
# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2019-03-15 10:52+0100\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
```
All general terms in the header should be replaced by specific information.

In the body, directly following the header, for each descriptive text there is an entry of the form
```
#: <file name>:<line number>
msgctxt "<context identifier>"
msgid "<descriptive text>"
msgstr "<translation>"
```
Only the keywords `msgctxt`, `msgid` and `msgstr` are used.

At first there can be an optional comment describing the location (file name and line number) of the text to translate. Multiple occurences of the same string can be listed here, separated by space.

> Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes belong to the enclosing full class definition.

Reviewer 4: Short class definitions can have *description*, according to [the grammar](https://specification.modelica.org/master/modelica-concrete-syntax.html#class-definition), which is *description-string* and *annotation*. Shouldn't they be translated too?

Reviewer: Yes, but they are confusingly referred to as "syntactical comments" in this document.

Reviewer 2: And note that the text (now?) makes it clear that those texts will be part of the enclosing class.

> Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes belong to the enclosing full class definition.

Reviewer 2: ```suggestion
Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes (including their description string) belong to the enclosing full class definition.
```
